lib/SILGen/SILGenPoly.cpp

```diff
@@ -294,6 +295,90 @@ SILGenFunction::emitTransformExistential(SILLocation loc,
+  // Look up the `Optional<T>.TangentVector.init` declaration.
+  auto initLookup =
+    optionalTanDecl->lookupDirect(DeclBaseName::createConstructor());
+  ConstructorDecl *constructorDecl = nullptr;
+  for (auto *candidate : initLookup) {
+    auto candidateModule = candidate->getModuleContext();
+    if (candidateModule->getName() ==
+        getASTContext().Id_Differentiation ||
+        candidateModule->isStdlibModule()) {
+      assert(!constructorDecl && "Multiple `Optional.TangentVector.init`s");
+      constructorDecl = cast<ConstructorDecl>(candidate);
+#ifdef NDEBUG
+      break;
+#endif
+    }
+  }
+  assert(constructorDecl && "No `Optional.TangentVector.init`");
+
+  // `Optional<T.TangentVector>`
+  CanType optionalOfWrappedTanType = inputType.wrapInOptionalType();
+
+  const TypeLowering &optTL = getTypeLowering(optionalOfWrappedTanType);
+  auto optVal = emitInjectOptional(loc, optTL, ctxt,
+                                   [&](SGFContext objectCtxt) {
+                                     return input;
+                                   });
+  auto *diffProto = getASTContext().getProtocol(KnownProtocolKind::Differentiable);
+  auto diffConf = lookupConformance(inputType, diffProto);
+  assert(!diffConf.isInvalid() && "Missing conformance to `Differentiable`");
+  ConcreteDeclRef initDecl(constructorDecl,
+                           SubstitutionMap::get(constructorDecl->getGenericSignature(),
+                                                {inputType}, {diffConf}));
+  PreparedArguments args({AnyFunctionType::Param(optionalOfWrappedTanType)});
+  args.add(loc, RValue(*this, {optVal}, optionalOfWrappedTanType));
+
+  auto result = emitApplyAllocatingInitializer(loc, initDecl,
+                                               std::move(args), outputType, ctxt);
+  if (result.isInContext())
+    return ManagedValue::forInContext();
+  return std::move(result).getAsSingleValue(*this, loc);
+}
+
+ManagedValue
+SILGenFunction::emitOptionalTangentVectorToTangentVector(SILLocation loc,
+                                                         ManagedValue input,
+                                                         CanType inputType,
+                                                         CanType outputType,
+                                                         SGFContext ctxt) {
+  // Optional<T>.TangentVector should be a struct with a single
+  // Optional<T.TangentVector> property. This is an implementation detail of
+  // OptionalDifferentiation.swift
+  // TODO: Maybe it would be better to have getters / setters here that we
+  // can call and hide this implementation detail?
+  StructDecl *optStructDecl = inputType.getStructOrBoundGenericStruct();
+  VarDecl *wrappedValueVar = nullptr;
+  if (optStructDecl) {
+    ArrayRef<VarDecl *> properties = optStructDecl->getStoredProperties();
+    wrappedValueVar = properties.size() == 1 ? properties[0] : nullptr;
+  }
```

Comment:
I think this would also be worth factoring out, looks copied from `PullbackCloner::Implementation::visitEnumInst`